siliconcompiler/core.py

```diff
@@ -4987,6 +5037,130 @@ def __getstate__(self):
+                self.set(*keypath, include_libraries, field='copy')
+
+        # Copy pdk files
+        for keypath in self.allkeys('pdk'):
+            if 'default' in keypath:
+                continue
+            keypath = ['pdk', *keypath]
+            key_type = self.get(*keypath, field='type')
+            if 'file' in key_type or 'dir' in key_type:
+                self.set(*keypath, include_pdks, field='copy')
+
+        collect_path = os.path.join(issue_dir.name, 'collect')
+        self._collect(directory=collect_path)
+        additional_files[collect_path] = os.path.join(self.get('option','builddir'),
+                                                        self.get('design'),
+                                                        self.get('option', 'jobname'),
+                                                        'sc_collected_files')
+
+        git_data = {}
+        try:
+            # Check git information
+            repo = git.Repo(path=os.path.join(self.scroot, '..'))
+            git_data['commit'] = repo.head.commit.hexsha
+            git_data['date'] = time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime(repo.head.commit.committed_date))
+            git_data['author'] = f'{repo.head.commit.author.name} <{repo.head.commit.author.email}>'
+            git_data['msg'] = repo.head.commit.message
+            # Count number of commits ahead of version
+            version_tag = repo.tag(f'v{self.scversion}')
+            count = 0
+            for c in repo.head.commit.iter_parents():
+                count += 1
+                if c == version_tag.commit:
+                    break
+            git_data['count'] = count
+        except git.InvalidGitRepositoryError:
+            pass
+
+        issue_information = {}
+        issue_information['environment'] = {key: value for key, value in os.environ.items()}
+        issue_information['python'] = {"path": sys.path,
+                                       "version": sys.version}
+        issue_information['date'] = datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d %H:%M:%S')
+        issue_information['machine'] = Chip._get_machine_info()
+        issue_information['run'] = {'step': step,
+                                    'index': index,
+                                    'libraries_included': include_libraries,
+                                    'pdks_included': include_pdks}
+        issue_information['version'] = {'schema': self.schemaversion,
+                                        'sc': self.scversion,
+                                        'git': git_data}
+
+        issue_path = os.path.join(issue_dir.name, 'issue.json')
+        with open(issue_path, 'w') as fd:
+            json.dump(issue_information, fd, indent=4, sort_keys=True)
+        additional_files[issue_path] = 'issue.json'
+
+        self.archive(nodelist=[(step, index)], all_files=True, include_manifest=False, archive_name=archive_name, additional_files=additional_files)
```

Comment:
I wonder if it would make sense to factor out some of `archive()`'s functionality into a private method that gets used here, rather than adding parameters to `archive()` itself. I'm thinking something along the lines of an `_archive_node()` that takes in a `Tarfile`, `step`, and `index`, and implements the `archive()` inner-loop logic. The `Tarfile` could be instantiated by the caller, and they could add any additional files directly to that. 

I think it would help keep the public API a bit cleaner, and I had this reaction after realizing that the additional files passed in here don't necessarily come from the job directory (whereas when I first read the API changes I assumed `archive()` would be searching for `additional_files` relative to the job being archived).

Reply:
refactored